### PR TITLE
Implement permissions info api backend

### PR DESCRIPTION
**Pull request by thenamankumar**
Implement API for list which user has access to which repository and why.

closes 50% of #47869 

TODO:
- [ ] Write unit tests
- [ ] Add more comments in the code
- [ ]  fix 1 bug
 
## Test plan

image

### CODE REVIEW

File: internal/database/repos.go [hunk 1]
```diff
@@ -188,7 +188,7 @@ func logPrivateRepoAccessGranted(ctx context.Context, db DB, ids []api.RepoID) {
 		event.AnonymousUserID = "internal"
 	}
 
-	db.SecurityEventLogs().LogEvent(ctx, event)
+	//db.SecurityEventLogs().LogEvent(ctx, event)
```
Comment (thenamankumar) [inline internal/database/repos.go:191]: NEED HELP!!!

This line causes In-valid pointer reference panic. Unable to fix this. Need help.

File: enterprise/internal/database/perms_store.go [hunk 2]
```diff
@@ -2042,3 +2046,149 @@ func sliceToSet[T comparable](s []T) map[T]struct{} {
...
+	UserRepoPermissionReasonUnrestricted   UserRepoPermissionReason = "Unrestricted"
+	UserRepoPermissionReasonPermissionSync UserRepoPermissionReason = "Permission Sync"
+)
+
+type ListUserPermissionsArgs struct {
+	Query          string
+	PaginationArgs *database.PaginationArgs
+}
+
+func (s *permsStore) getRepoPermissionInfoForIDs(ctx context.Context, userID int32, repoIDs []int32) (map[int32]*repoPermissionInfo, error) {
```
Comment (thenamankumar) [inline enterprise/internal/database/perms_store.go:2139]: Integration with the new data model will be done under a separate PR.

File: cmd/frontend/graphqlbackend/authz.graphql [hunk 3]
```diff
@@ -293,7 +293,7 @@ type PermissionsInfo {
     """
     The permission levels that a user has on the repository.
     """
-    permissions: [RepositoryPermission!]!
+    permissions: [RepositoryPermission!]! @deprecated(reason: "we only ever use READ permissions.")
```
Comment (thenamankumar) [inline cmd/frontend/graphqlbackend/authz.graphql:296]: better message required.

File: enterprise/cmd/frontend/internal/authz/resolvers/resolver.go [hunk 4]
```diff
@@ -544,6 +552,91 @@ func (r *permissionsInfoResolver) Unrestricted() bool {
...
+
+	return &cursor, nil
+}
+
+func (s *permissionsInfoRepositoriesStore) UnmarshalCursor(cursor string, _ database.OrderBy) (*string, error) {
+	return &cursor, nil
+}
+
+func (s *permissionsInfoRepositoriesStore) ComputeTotal(ctx context.Context) (*int32, error) {
+	total := int32(20)
```
Comment (thenamankumar) [inline enterprise/cmd/frontend/internal/authz/resolvers/resolver.go:601]: this is a todo

File: enterprise/internal/database/perms_store.go [hunk 5]
```diff
@@ -2042,3 +2046,149 @@ func sliceToSet[T comparable](s []T) map[T]struct{} {
 	}
 	return m
 }
+
+// ListUserPermissions gets the list of accessible repos for the provided user, along with the reason
+// and timestamp for each permission.
+func (s *permsStore) ListUserPermissions(ctx context.Context, userID int32, args *ListUserPermissionsArgs) (perms []*UserPermission, err error) {
```
Comment (sashaostrikov) [inline enterprise/internal/database/perms_store.go:2052]: I would make this function work other way around. The contract is to return one page (e.g. of size 20) of repos which can be accessed by user.

#### Currently
We fetch 20 repos -> we check permissions (I believe in not a bug-free way) -> we return repos.

There is a possibility (however, not with current code) that we will fetch repos which cannot be accessed by a given user. Then we need to filter this repos and we will end up with less than 20 repos returned.

To fix this I think we should...

#### Do this differently
Support pagination in `getRepoPermissionInfoForIDs` and fetch 20 repos which we know for sure can be accessed by the user. I believe we can support repo name querying as well because we anyway join with `repo` table.

WDYT?

File: internal/database/repos.go (hunk 1 above)
Comment (sashaostrikov) [reply internal/database/repos.go:191]: Sorry, too little context. If you can at least provide the stack trace.
It works on main, I believe maybe there is something tricky with the way you create a `db` in `perms_store` or something like that which doesn't contain an implementation of `SecurityEventLogs`...?

File: internal/database/repos_perm.go [hunk 6]
```diff
@@ -14,23 +14,35 @@ import (
...
-func AuthzQueryConds(ctx context.Context, db DB) (*sqlf.Query, error) {
+type BypassAuthzReason = string
+
+const (
+	BypassAuthzReasonSiteAdmin       BypassAuthzReason = "Site Admin"
+	BypassAuthzReasonIsInternal      BypassAuthzReason = "Internal Request"
+	BypassAuthzReasonNoAuthzProvider BypassAuthzReason = "No Authz Provider Configured"
+)
+
+type AuthzQueryParameters struct {
```
Comment (sashaostrikov) [inline internal/database/repos_perm.go:25]: I guess we can even add a nice `authzQuery` method to this struct

**Review by sashaostrikov — COMMENTED** [review]
Nice progress!
I'd note a couple of things here:
1) This PR can certainly be many smaller PRs, I would begin with adding a _well-tested_ function to list repos the user has access to.
2) Make any change related to this PR in a TDD manner, because it is a very tricky and bug-prone area of code. Write a lot of non-abstract test cases, you can even use prod/s2 data for entries in repo/user permissions table.

File: cmd/frontend/graphqlbackend/authz.graphql [hunk 7]
```diff
@@ -309,6 +309,70 @@ type PermissionsInfo {
     It will always be false for users.
     """
     unrestricted: Boolean!
+
+    """
+    The connection for repository permissions avilable to the user. It will return
+    no results for repo parent.
```
[the remark this answers is not in the document]
Comment (kopancek) [reply cmd/frontend/graphqlbackend/authz.graphql:315]: Same

File: cmd/frontend/graphqlbackend/authz.graphql [hunk 8]
```diff
@@ -309,6 +309,70 @@ type PermissionsInfo {
...
+    """
+    repositories(
```
Comment (kopancek) [inline cmd/frontend/graphqlbackend/authz.graphql:317]: This only makes sense for users. Should we instead create another type for user permissions? It might be better than putting both into the PermissionsInfo struct

File: enterprise/cmd/frontend/internal/authz/resolvers/resolver.go [hunk 9]
```diff
@@ -519,12 +522,17 @@ func getOrDefault[T any](ptr *T) T {
...
+	ossDB        database.DB
+	userID       int32
+	repoID       int32
 	perms        authz.Perms
 	syncedAt     time.Time
 	updatedAt    time.Time
 	unrestricted bool
 }
 
+// NOTE(naman): deprecated as r.perms is always "READ", making the info redundant.
```
Comment (kopancek) [inline enterprise/cmd/frontend/internal/authz/resolvers/resolver.go:535]: Instead of a note, can we use GoLang's [standard way to deprecate functions](https://rakyll.org/deprecated/)?

File: enterprise/cmd/frontend/internal/authz/resolvers/resolver.go [hunk 10]
```diff
@@ -544,6 +552,91 @@ func (r *permissionsInfoResolver) Unrestricted() bool {
...
+var permissionsInfoRepositoryConnectionOptions = &graphqlutil.ConnectionResolverOptions{
+	OrderBy:     database.OrderBy{{Field: "repo.name"}},
+	Ascending:   true,
+	MaxPageSize: &permissionsInfoRepositoryConnectionMaxPageSize,
+}
+
+func (r *permissionsInfoResolver) Repositories(ctx context.Context, args graphqlbackend.PermissionsInfoRepositoriesArgs) (*graphqlutil.ConnectionResolver[graphqlbackend.PermissionsInfoRepositoryResolver], error) {
+	if r.userID == 0 {
+		return nil, nil
+	}
```
Comment (kopancek) [inline enterprise/cmd/frontend/internal/authz/resolvers/resolver.go:566]: Should we return an error instead saying userID is invalid?

File: enterprise/cmd/frontend/internal/authz/resolvers/resolver.go (hunk 4 above)
Comment (kopancek) [reply enterprise/cmd/frontend/internal/authz/resolvers/resolver.go:601]: Can you mark it as such in the code maybe? :) Would make it harder to forget in comparison to a comment on a PR

File: enterprise/cmd/frontend/internal/authz/resolvers/resolver.go [hunk 11]
```diff
@@ -544,6 +552,91 @@ func (r *permissionsInfoResolver) Unrestricted() bool {
...
+		userID: r.userID,
+		db:     r.db,
+		ossDB:  r.ossDB,
+		query:  query,
+	}
+
+	return graphqlutil.NewConnectionResolver[graphqlbackend.PermissionsInfoRepositoryResolver](connectionStore, &args.ConnectionResolverArgs, permissionsInfoRepositoryConnectionOptions)
+}
+
+type permissionsInfoRepositoriesStore struct {
```
Comment (kopancek) [inline enterprise/cmd/frontend/internal/authz/resolvers/resolver.go:583]: Can we use less complex naming here? E.g.
```suggestion
type repositoryPermissionsStore struct {
```

File: enterprise/cmd/frontend/internal/authz/resolvers/resolver.go [hunk 12]
```diff
@@ -544,6 +552,91 @@ func (r *permissionsInfoResolver) Unrestricted() bool {
...
+func (s *permissionsInfoRepositoriesStore) UnmarshalCursor(cursor string, _ database.OrderBy) (*string, error) {
+	return &cursor, nil
+}
+
+func (s *permissionsInfoRepositoriesStore) ComputeTotal(ctx context.Context) (*int32, error) {
+	total := int32(20)
+	return &total, nil
+}
+
+func (s *permissionsInfoRepositoriesStore) ComputeNodes(ctx context.Context, args *database.PaginationArgs) ([]graphqlbackend.PermissionsInfoRepositoryResolver, error) {
```
Comment (kopancek) [inline enterprise/cmd/frontend/internal/authz/resolvers/resolver.go:605]:
```suggestion
func (s *permissionsInfoRepositoriesStore) ComputeNodes(ctx context.Context, args *database.PaginationArgs) ([]graphqlbackend.RepositoryPermissionsResolver, error) {
```

File: enterprise/cmd/frontend/internal/authz/resolvers/resolver.go [hunk 13]
```diff
@@ -544,6 +552,91 @@ func (r *permissionsInfoResolver) Unrestricted() bool {
...
+	permissions, err := s.db.Perms().ListUserPermissions(ctx, s.userID, &edb.ListUserPermissionsArgs{Query: s.query, PaginationArgs: args})
+	if err != nil {
+		return nil, err
+	}
+
+	var permissionResolvers []graphqlbackend.PermissionsInfoRepositoryResolver
```
Comment (kopancek) [inline enterprise/cmd/frontend/internal/authz/resolvers/resolver.go:611]: We know the size upfront, so might initialize it as such:
```suggestion
	permissionResolvers := make([]graphqlbackend.PermissionsInfoRepositoryResolver, 0, len(permissions))
```

File: enterprise/cmd/frontend/internal/authz/resolvers/resolver.go [hunk 14]
```diff
@@ -544,6 +552,91 @@ func (r *permissionsInfoResolver) Unrestricted() bool {
...
+	for _, perm := range permissions {
+		permissionResolvers = append(permissionResolvers, permissionsInfoRepositoryResolver{perm: perm, ossDB: s.ossDB})
+	}
+
+	return permissionResolvers, nil
+}
+
+type permissionsInfoRepositoryResolver struct {
+	ossDB database.DB
```
Comment (kopancek) [inline enterprise/cmd/frontend/internal/authz/resolvers/resolver.go:620]: No need to distinguish here, so might as well call it `db`?

```suggestion
	db database.DB
```

File: enterprise/cmd/frontend/internal/authz/resolvers/resolver.go [hunk 15]
```diff
@@ -519,12 +522,17 @@ func getOrDefault[T any](ptr *T) T {
 }
 
 type permissionsInfoResolver struct {
+	db           edb.EnterpriseDB
+	ossDB        database.DB
```
Comment (kopancek) [inline enterprise/cmd/frontend/internal/authz/resolvers/resolver.go:526]: Consider renaming `db` to something better, since we apparently need both:
```suggestion
	eDB           edb.EnterpriseDB
	ossDB        database.DB
```

File: enterprise/internal/database/perms_store.go [hunk 16]
```diff
@@ -2042,3 +2046,149 @@ func sliceToSet[T comparable](s []T) map[T]struct{} {
...
+	ctx, save := s.observe(ctx, "ListUserPermissions", "")
+	defer save(&err)
+
+	// Set actor with provided userID to context.
+	ctx = actor.WithActor(ctx, actor.FromUser(userID))
+
+	db := database.NewDBWith(s.logger, s.Store)
+
+	// NOTE(naman): `List` uses the actor from ctx to apply authz conds
+	// to fetch list of repos which are visible to the actor.
```
Comment (kopancek) [inline enterprise/internal/database/perms_store.go:2062]: Let's not say `visible`, but `accessible` here. Since gitlab incident, I'd like to be specific in our language:
```suggestion
	// to fetch list of repos which are accessible by the actor.
```

File: enterprise/internal/database/perms_store.go [hunk 17]
```diff
@@ -2042,3 +2046,149 @@ func sliceToSet[T comparable](s []T) map[T]struct{} {
...
+	if authzParams.BypassAuthz {
+		// NOTE(naman): If Authz conds were bypassed then no need to fetch
+		// perm details for the repos. Show permission reason as "Site Admin"
+		// or "Authentication Bypassed" based on query params used.
+
+		reason := UserRepoPermissionReasonAuthzBypass
+
+		if authzParams.BypassAuthzReason == database.BypassAuthzReasonSiteAdmin {
+			reason = UserRepoPermissionReasonSiteAdmin
+		}
```
Comment (kopancek) [inline enterprise/internal/database/perms_store.go:2082]: Would this code better fit into the `getRepoPermissionInfoForIDs` function?

File: enterprise/internal/database/perms_store.go [hunk 18]
```diff
@@ -2042,3 +2046,149 @@ func sliceToSet[T comparable](s []T) map[T]struct{} {
...
+
+		for _, repo := range repos {
+			perms = append(perms, &UserPermission{Repo: repo, Reason: reason})
+		}
+	} else {
+		// NOTE(naman): If Authz conds were not bypassed then find perm details
+		// for each of the repos and show correct reason and updatedAt.
+
+		repoIDs := []int32{}
```
Comment (kopancek) [inline enterprise/internal/database/perms_store.go:2091]: We know the size already. Initializing upfront will save us a few memory allocations:
```suggestion
		repoIDs := make([]int32, 0, len(repos))
```

File: enterprise/internal/database/perms_store.go [hunk 19]
```diff
@@ -2042,3 +2046,149 @@ func sliceToSet[T comparable](s []T) map[T]struct{} {
...
+		for _, repo := range repos {
+			repoIDs = append(repoIDs, int32(repo.ID))
+		}
```
Comment (kopancek) [inline enterprise/internal/database/perms_store.go:2094]: Do we need this type casting? Can we use `api.RepoID` everywhere instead?

File: enterprise/internal/database/perms_store.go [hunk 20]
```diff
@@ -2042,3 +2046,149 @@ func sliceToSet[T comparable](s []T) map[T]struct{} {
...
+	ctx = actor.WithActor(ctx, actor.FromUser(userID))
+
+	db := database.NewDBWith(s.logger, s.Store)
+
+	// NOTE(naman): `List` uses the actor from ctx to apply authz conds
+	// to fetch list of repos which are visible to the actor.
+	repos, err := db.Repos().List(ctx, database.ReposListOptions{
+		PaginationArgs: args.PaginationArgs,
+		Query:          args.Query,
+	})
```
Comment (kopancek) [inline enterprise/internal/database/perms_store.go:2066]: Should we run this in a transaction? What happens if permissions are updated between getting repos here and checking the permissions table below?

File: enterprise/internal/database/perms_store.go (hunk 5 above)
Comment (kopancek) [reply enterprise/internal/database/perms_store.go:2052]: I like this approach more as well, but paginating it might be hard.

File: enterprise/internal/database/perms_store.go [hunk 21]
```diff
@@ -2042,3 +2046,149 @@ func sliceToSet[T comparable](s []T) map[T]struct{} {
...
+// ListUserPermissions gets the list of accessible repos for the provided user, along with the reason
+// and timestamp for each permission.
+func (s *permsStore) ListUserPermissions(ctx context.Context, userID int32, args *ListUserPermissionsArgs) (perms []*UserPermission, err error) {
+	ctx, save := s.observe(ctx, "ListUserPermissions", "")
+	defer save(&err)
+
+	// Set actor with provided userID to context.
+	ctx = actor.WithActor(ctx, actor.FromUser(userID))
+
+	db := database.NewDBWith(s.logger, s.Store)
```
[the remark this answers is not in the document]
Comment (kopancek) [reply enterprise/internal/database/perms_store.go:2059]: It's already on the constructor, so I don't see why not.

File: enterprise/internal/database/perms_store.go [hunk 22]
```diff
@@ -2042,3 +2046,149 @@ func sliceToSet[T comparable](s []T) map[T]struct{} {
...
+		WHEN up.user_id IS NOT NULL THEN 'Permission Sync'
+		ELSE 'Unrestricted'
+	END AS permission_reason
+FROM
+	repo
+	LEFT JOIN user_permissions AS up ON up.user_id = %d
+		AND up.object_ids_ints @> INTSET (repo.id)
+WHERE
+	repo.id = ANY (%s)
+`
```
Comment (kopancek) [inline enterprise/internal/database/perms_store.go:2174]: The query is not looking at the external service, which might also be marked as `unrestricted` in case permissions are not enabled. 

I think this query would give us what we want:
```suggestion
-- restricted repos
SELECT unnest(object_ids_ints) AS id, 'Permission Sync' AS permission_reason
  FROM user_permissions AS up
  WHERE up.user_id  = %d
UNION
-- unrestricted repos
SELECT repo_id AS id, 'Unrestricted' AS permission_reason
  FROM repo_permissions
  WHERE unrestricted
`

And we need to do a different query if `globals.PermissionsUserMapping().Enabled` is `false`, like we do in `authzQuery`.

And we can do a completely different query for site admins if their permissions are not enforced (e.g. just paginate through repos and give `site_admin` as a reason).
```

File: internal/database/repos_perm.go [hunk 23]
```diff
@@ -40,19 +52,40 @@ func AuthzQueryConds(ctx context.Context, db DB) (*sqlf.Query, error) {
...
 	// 🚨 SECURITY: internal requests bypass authz provider permissions checks,
 	// so correctness is important here.
-	bypassAuthz := a.IsInternal() || (authzAllowByDefault && len(authzProviders) == 0)
-	if !bypassAuthz && a.IsAuthenticated() {
+	if a.IsInternal() {
+		params.BypassAuthz = true
+		params.BypassAuthzReason = BypassAuthzReasonIsInternal
+	} else if authzAllowByDefault && len(authzProviders) == 0 {
+		params.BypassAuthz = true
+		params.BypassAuthzReason = BypassAuthzReasonNoAuthzProvider
```
Comment (kopancek) [inline internal/database/repos_perm.go:60]:
```suggestion
		params.BypassAuthzReason = BypassAuthzAllowByDefault
```

File: internal/database/repos_perm.go [hunk 24]
```diff
@@ -40,19 +52,40 @@ func AuthzQueryConds(ctx context.Context, db DB) (*sqlf.Query, error) {
...
+	if err != nil {
+		return nil, err
 	}
 
-	q := authzQuery(bypassAuthz,
-		usePermissionsUserMapping,
-		authenticatedUserID,
+	q := authzQuery(params.BypassAuthz,
+		params.UsePermissionsUserMapping,
+		params.AuthenticatedUserID,
```
Comment (kopancek) [inline internal/database/repos_perm.go:88]: I think since we already have it in object, let's simplify?

```suggestion
	q := authzQuery(params,
```

File: internal/database/repos.go (hunk 1 above)
Comment (thenamankumar) [reply internal/database/repos.go:191]: Its most probably how I create db in the perms_store. Better to resolve it sync on Monday I guess.

File: enterprise/cmd/frontend/internal/authz/resolvers/resolver.go (hunk 4 above)
Comment (thenamankumar) [reply enterprise/cmd/frontend/internal/authz/resolvers/resolver.go:601]: I'll resolve this before submitting it for real review.

File: cmd/frontend/graphqlbackend/authz.graphql (hunk 7 above)
[the remark this answers is not in the document]
Comment (thenamankumar) [reply cmd/frontend/graphqlbackend/authz.graphql:315]: gql naming. 

Basically `PermissionsInfo` type works for both `user.permissionsInfo` and `repo.permissionInfo`, with user and repo being parent types, respectively. 

So it will return no results for `repo.permissionsInfo`.


Anyways I'll update the description with something better.

File: enterprise/cmd/frontend/internal/authz/resolvers/resolver.go [hunk 25]
```diff
@@ -519,12 +522,17 @@ func getOrDefault[T any](ptr *T) T {
...
+	userID       int32
+	repoID       int32
```
[the remark this answers is not in the document]
Comment (thenamankumar) [reply enterprise/cmd/frontend/internal/authz/resolvers/resolver.go:528]: why not, I'll try with that. thanks .

File: enterprise/internal/database/perms_store.go [hunk 26]
```diff
@@ -226,6 +228,8 @@ type PermsStore interface {
 	// It filters out empty bindIDs and only returns users that exist in the database.
 	// If a bind id doesn't map to any user, it is ignored.
 	MapUsers(ctx context.Context, bindIDs []string, mapping *schema.PermissionsUserMapping) (map[string]int32, error)
+	// ListUserPermissions returns list of repository permissions info the user has access to.
+	ListUserPermissions(ctx context.Context, userID int32, args *ListUserPermissionsArgs) (perms []*UserPermission, err error)
```
[the remark this answers is not in the document]
Comment (thenamankumar) [reply enterprise/internal/database/perms_store.go:232]: yeahhhhhh

File: enterprise/internal/database/perms_store.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (thenamankumar) [reply enterprise/internal/database/perms_store.go:2059]: @kopancek, the db passed to the constructer is again edb. I'll recheck and try to modify the constructor if required.

[the remark this answers is not in the document]
Comment (thenamankumar) [reply enterprise/internal/database/perms_store.go:2059]: Definitely need to fix this as I have a runtime error with logger when using ossDB from perms_store right now.

File: enterprise/internal/database/perms_store.go (hunk 5 above)
Comment (thenamankumar) [reply enterprise/internal/database/perms_store.go:2052]: The pagination won't be an issue. 

I initially wanted one single query. But merging the two into one means we can not reuse the Repos().LIst() we have right now. And that will require copying the repo row scanner code which is long. 

I got greedy with writing short, more manageable code. 

But yes, it might cause unexpected results for certain edge cases. So I'll try to construct a single SQL query with pagination using the AuthzConds directly, which should resolve the concern. 

wdyt?

File: enterprise/internal/database/perms_store.go [hunk 27]
```diff
@@ -2042,3 +2046,149 @@ func sliceToSet[T comparable](s []T) map[T]struct{} {
...
+			reason := UserRepoPermissionReasonUnrestricted
+
+			info := permissionsInfoMap[int32(repo.ID)]
+			if info != nil {
+				reason = info.Reason
+				updatedAt = info.UpdatedAt
+			}
+
+			perms = append(perms, &UserPermission{Repo: repo, Reason: reason, UpdatedAt: updatedAt})
+		}
```
[the remark this answers is not in the document]
Comment (thenamankumar) [reply enterprise/internal/database/perms_store.go:2112]: I don't get where the user 2 comes from inside `ListUserPermissions`. There is only 1 user we care about and that comes from the `userID` arg.

File: enterprise/internal/database/perms_store.go [hunk 28]
```diff
@@ -2042,3 +2046,149 @@ func sliceToSet[T comparable](s []T) map[T]struct{} {
...
+			return infoMap, err
+		}
+
+		infoMap[info.RepoID] = info
+	}
+
+	return infoMap, nil
+}
+
+const permissionInfoQueryFmt = `
```
[the remark this answers is not in the document]
Comment (thenamankumar) [reply enterprise/internal/database/perms_store.go:2160]: All those cases are handeled with `Repos().List()` query. 

We first fetch the list of repos accessible to the users using Repos().List(). 20 of them. 

And only then we check if the user has access because of permissions sync or being unrestricted for those 20 repos specifically.

File: enterprise/internal/database/perms_store.go (hunk 22 above)
Comment (thenamankumar) [reply enterprise/internal/database/perms_store.go:2174]: All of that is handled by the `Repos().List()`.

File: enterprise/internal/database/perms_store.go [hunk 29]
```diff
@@ -2042,3 +2046,149 @@ func sliceToSet[T comparable](s []T) map[T]struct{} {
...
+
+	infoMap := make(map[int32]*repoPermissionInfo, len(repoIDs))
+	for rows.Next() {
+		info, err := scanRepoPermInfo(rows)
+		if err != nil {
+			return infoMap, err
+		}
+
+		infoMap[info.RepoID] = info
+	}
```
Comment (mrnugget) [inline enterprise/internal/database/perms_store.go:2155]: Maybe `NewMapScanner` can help? https://sourcegraph.sourcegraph.com/search?q=context%3Asourcegraph+NewMapScanner(&groupBy=repo
